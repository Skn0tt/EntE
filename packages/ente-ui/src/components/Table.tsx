import * as React from "react";
import MUIDataTable from "mui-datatables";
import * as _ from "lodash";
import { LanguageProvider } from "../helpers/Language-Provider";
import { getByLanguage } from "ente-types";

export const translation = getByLanguage({
  en: {
    textLabels: {
      body: {
        noMatch: "Sorry, no matching records found",
        toolTip: "Sort"
      },
      pagination: {
        next: "Next Page",
        previous: "Previous Page",
        rowsPerPage: "Rows per page:",
        displayRows: "of"
      },
      toolbar: {
        search: "Search",
        downloadCsv: "Download CSV",
        print: "Print",
        viewColumns: "View Columns",
        filterTable: "Filter Table"
      },
      filter: {
        all: "All",
        title: "FILTERS",
        reset: "RESET"
      },
      viewColumns: {
        title: "Show Columns",
        titleAria: "Show/Hide Table Columns"
      },
      selectedRows: {
        text: "rows(s) selected",
        delete: "Delete",
        deleteAria: "Delete Selected Rows"
      }
    }
  },
  de: {
    textLabels: {
      body: {
        noMatch: "Es liegen keine Datensätze vor.",
        toolTip: "Sortieren"
      },
      pagination: {
        next: "Nächste Seite",
        previous: "Vorherige Seite",
        rowsPerPage: "Zeilen pro Seite:",
        displayRows: "von"
      },
      toolbar: {
        search: "Suchen",
        downloadCsv: "Als CSV herunterladen",
        print: "Drucken",
        viewColumns: "Spalten zeigen",
        filterTable: "Tabelle filtern"
      },
      filter: {
        all: "Alles",
        title: "Filter".toUpperCase(),
        reset: "Zurücksetzen".toUpperCase()
      },
      viewColumns: {
        title: "Spalten",
        titleAria: "Tabellenzeilen Zeigen/Verstecken"
      },
      selectedRows: {
        text: "Zeile(n) ausgewählt",
        delete: "Löschen",
        deleteAria: "Ausgewählte Zeilen löschen"
      }
    }
  }
});

interface ColumnConfig<S, T> {
  name: string;
  extract: (v: S) => T;
  options?: {
    display?: boolean;
    filter?: boolean;
    sort?: boolean;
    customBodyRender?: (v: T) => string | JSX.Element;
  };
}

interface TableOwnProps<T> {
  items: ReadonlyArray<T>;
  columns: ReadonlyArray<ColumnConfig<T, any>>;
  extractId: (item: T) => string;
  onClick?: (id: string) => void;
  title?: string | JSX.Element;
}

type TableProps<T> = TableOwnProps<T>;

export class Table<T> extends React.PureComponent<TableProps<T>> {
  render() {
    const { columns, items, extractId, onClick = () => {}, title } = this.props;

    const data = items.map(item => {
      const id = extractId(item);
      const cells = columns.map(columnOptions => {
        return columnOptions.extract(item);
      });
      return [id, ...cells];
    });
    const idColumn = { name: "ID", options: { display: false, filter: false } };

    return (
      <LanguageProvider>
        {lang => (
          <MUIDataTable
            columns={[idColumn, ...columns]}
            data={data}
            title={title as any}
            options={{
              rowsPerPage: 50,
              rowsPerPageOptions: [20, 50, 100],
              selectableRows: false,
              elevation: 1,
              responsive: "scroll",
              onRowClick: (d: any[]) => onClick(d[0]),
              textLabels: translation(lang).textLabels
            }}
          />
        )}
      </LanguageProvider>
    );
  }
}

export default Table;
