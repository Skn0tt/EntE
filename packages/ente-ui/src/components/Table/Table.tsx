import * as React from "react";
import MUIDataTable from "mui-datatables";
import * as _ from "lodash";

interface ColumnConfig {
  name: string;
  options?: {
    display?: boolean;
    filter?: boolean;
    sort?: boolean;
    customBodyRender?: (v: string) => string | JSX.Element;
  };
}

type ConfigItem = ColumnConfig | string;

interface TableProps<T> {
  items: ReadonlyArray<T>;
  extract: (item: T) => string[];
  headers: ReadonlyArray<ConfigItem>;
  extractId: (item: T) => string;
  onClick?: (id: string) => void;
}

export class Table<T> extends React.PureComponent<TableProps<T>> {
  render() {
    const { headers, items, extract, extractId, onClick } = this.props;
    const data = items.map(i => [extractId(i), ...extract(i)]);
    const columns = headers.map(
      h => (typeof h === "string" ? { name: h, options: { filter: false } } : h)
    );
    const idColumn = { name: "ID", options: { display: false, filter: false } };
    return (
      <MUIDataTable
        columns={[idColumn, ...columns]}
        data={data}
        options={{
          rowsPerPage: 50,
          rowsPerPageOptions: [20, 50, 100],
          selectableRows: false,
          responsive: "scroll",
          onRowClick: d => onClick(d[0])
        }}
      />
    );
  }
}

export default Table;
