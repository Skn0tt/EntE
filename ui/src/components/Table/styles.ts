import { Theme, StyleRules } from 'material-ui/styles';

const styles = (theme: Theme): StyleRules => ({
  searchBar: {
    padding: 10,
  },
  fab: {
    margin: 0,
    top: 'auto',
    right: theme.spacing.unit * 2,
    bottom: theme.spacing.unit * 2,
    left: 'auto',
    position: 'fixed',
  },
  table: {
    overflowX: 'auto',
  },
});

export default styles;
