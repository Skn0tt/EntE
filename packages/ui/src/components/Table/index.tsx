import * as React from "react";
import styles from "./styles";
import withStyles from "material-ui/styles/withStyles";
import {
  TableCell,
  Grid,
  TextField,
  TableHead,
  TableRow,
  TableBody,
  WithStyles,
  Table as MUITable
} from "material-ui";
import TableHeadCell from "./elements/TableHeadCell";
import lang from "ente-lang";
import * as _ from "lodash";

/**
 * # Component Types
 */
type AllowedValues = string | number | boolean;
type Item = ReadonlyArray<AllowedValues>;

interface OwnProps<T> {
  defaultSortField?: number;
  items: ReadonlyArray<T>;
  cellExtractor: (item: T) => Item;
  headers: ReadonlyArray<string>;
  onClick?: (item: T) => void;
  keyExtractor: (item: T) => string | number;
  trueElement?: JSX.Element;
  falseElement?: JSX.Element;
  sort?: (a: T, b: T) => number;
  filter?: (item: T) => boolean;
}

type Props<T> = OwnProps<T> & WithStyles;

interface State {
  searchTerm: string;
  sortField: number;
  sortUp: boolean;
}

/**
 * # Logic
 */
export const changeSearch = (v: string) => (s: State): State => ({
  ...s,
  searchTerm: v
});

export const changeSortField = (i: number) => (s: State): State =>
  s.sortField === i
    ? { ...s, sortUp: !s.sortUp }
    : { ...s, sortField: i, sortUp: true };

/**
 * # Component
 */
export class Table<T> extends React.PureComponent<Props<T>, State> {
  /**
   * ## Intialization
   */
  state: State = {
    searchTerm: "",
    sortField: this.props.defaultSortField || 0,
    sortUp: false
  };

  /**
   * ## Handlers
   */
  handleChangeSearch = (event: React.ChangeEvent<HTMLInputElement>) =>
    this.setState(changeSearch(event.target.value));
  handleChangeSort = (i: number) => this.setState(changeSortField(i));

  /**
   * Logic
   */
  filter = (item: T): boolean => {
    const value = this.props.cellExtractor(item)[this.state.sortField];

    if (typeof value === "boolean") {
      return true;
    }

    return ("" + value).indexOf(this.state.searchTerm) !== -1;
  };

  sortMulti = () => (this.state.sortUp ? 1 : -1);
  sort = (a: T, b: T): number => {
    const valueA = this.props.cellExtractor(a)[this.state.sortField];
    const valueB = this.props.cellExtractor(b)[this.state.sortField];
    if (typeof valueA === "boolean") {
      return valueA === valueB ? 0 : valueA ? -1 : 1;
    }

    if (typeof valueA === "number") {
      return (valueA - (valueB as number)) * this.sortMulti();
    }

    return valueA.localeCompare(valueB as string) * this.sortMulti();
  };
  truncate = (str: string, length: number, ending: string) =>
    str.length > length
      ? str.substring(0, length - ending.length) + ending
      : str;
  show = (v: AllowedValues) => {
    switch (typeof v) {
      case "string":
        return _.truncate(v as string, { length: 20 });
      case "boolean":
        return v ? this.props.trueElement : this.props.falseElement;
      default:
        return v;
    }
  };

  render(): JSX.Element {
    const {
      classes,
      keyExtractor,
      onClick,
      items,
      headers,
      sort,
      filter,
      cellExtractor
    } = this.props;
    const { sortField, sortUp } = this.state;

    return (
      <Grid container direction="column">
        <Grid item container direction="row">
          <Grid item xs={12}>
            <TextField
              placeholder={lang().ui.table.search}
              fullWidth
              className={classes.searchBar}
              onChange={this.handleChangeSearch}
            />
          </Grid>
        </Grid>
        <Grid item className={classes.table}>
          <MUITable>
            <TableHead>
              <TableRow>
                {headers.map((header, i) => (
                  <TableHeadCell
                    key={header}
                    tooltip={lang().ui.table.sortTooltip(header)}
                    active={sortField === i}
                    sortUp={sortUp}
                    onClick={() => this.handleChangeSort(i)}
                  >
                    {header}
                  </TableHeadCell>
                ))}
              </TableRow>
            </TableHead>
            <TableBody>
              {items
                .filter(filter || this.filter)
                .sort(sort || this.sort)
                .map(item => (
                  <TableRow
                    key={keyExtractor(item)}
                    onClick={() => onClick && onClick(item)}
                    hover={!!onClick}
                  >
                    {cellExtractor(item).map((v, i) => (
                      <TableCell key={i}>{this.show(v)}</TableCell>
                    ))}
                  </TableRow>
                ))}
            </TableBody>
          </MUITable>
        </Grid>
      </Grid>
    );
  }
}

export default withStyles(styles)(Table);
