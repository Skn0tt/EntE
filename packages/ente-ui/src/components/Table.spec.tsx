/**
 * EntE
 * (c) 2017-present, Simon Knott <info@simonknott.de>
 *
 * This source code is licensed under the GNU Affero General Public License
 * found in the LICENSE file in the root directory of this source tree.
 */

import * as React from "react";
import { shallow } from "enzyme";
import { Table } from "./Table";

describe("Table", () => {
  type Person = {
    name: string;
    alter: number;
  };

  const items: Person[] = [
    {
      name: "Bernd",
      alter: 16
    },
    {
      name: "Volker",
      alter: 17
    }
  ];

  const comp = shallow(
    <Table<Person>
      columns={[
        { name: "Name", extract: v => v.name },
        { name: "Alter", extract: v => v.alter }
      ]}
      items={items}
      extractId={i => i.name}
    />
  );

  it("renders correctly", () => {
    expect(comp).toMatchSnapshot();
  });
});
