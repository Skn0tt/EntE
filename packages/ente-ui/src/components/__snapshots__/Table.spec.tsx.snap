// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Table
    columns={
      Array [
        Object {
          "extract": [Function],
          "name": "Name",
        },
        Object {
          "extract": [Function],
          "name": "Alter",
        },
      ]
    }
    extractId={[Function]}
    items={
      Array [
        Object {
          "alter": 16,
          "name": "Bernd",
        },
        Object {
          "alter": 17,
          "name": "Volker",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
